File: scripts/163.py
import math
from itertools import combinations
from tqdm import tqdm

# Global constant and helper
x = 1
TAN = lambda angle: round(math.tan(angle), 3)

# Precomputed allowed slopes for each label.
AXIS_SLOPES = {
    "CO": (0, None, TAN(math.pi/6), TAN(math.pi/3), TAN(2*math.pi/3), TAN(5*math.pi/6)),
    "C":  (None, TAN(math.pi/6), TAN(5*math.pi/6)),
    "TR": (TAN(math.pi/6), TAN(2*math.pi/3)),
    "BO": (0, None),
    "TL": (TAN(5*math.pi/6), TAN(math.pi/3)),
}

def are_same_axis(a, b):
    """
    Returns True if two points (each as ((x, y), label)) share an allowed axis.
    In the vertical case (x difference near 0), we check b’s label.
    Otherwise, we compute the slope (via atan→tan rounded to 3 decimals)
    and see if it is in the allowed set for a’s label.
    """
    (x_a, y_a), label_a = a
    (x_b, y_b), label_b = b
    if abs(x_a - x_b) < 1e-3:
        return label_b in ("CO", "C", "BO")
    slope = (y_a - y_b) / (x_a - x_b)
    tan_val = round(math.tan(math.atan(slope)), 3)
    return tan_val in AXIS_SLOPES.get(label_a, ())

def is_degenerate_triangle(a, b, c, tol=1e-6):
    """
    Uses the area formula to decide if the triangle defined by a, b, c (each as (x, y))
    is degenerate (i.e. nearly collinear or with coincident vertices).
    """
    if a == b or b == c or a == c:
        return True
    area = abs(a[0]*(b[1]-c[1]) + b[0]*(c[1]-a[1]) + c[0]*(a[1]-b[1])) / 2
    return area < tol

def gen_triangle(center, reversed=False):
    """
    Given a center point, returns a set of points (with labels) for a triangle.
    The triangle includes the center ("C") and six points around it, with orientation
    flipped if reversed==True.
    """
    R = math.pi if reversed else 0
    pts = { (center, "C") }
    for mult, offset, label in [
            (1, math.pi/6,    "TR"),
            (1, 5*math.pi/6,  "TL"),
            (1, 9*math.pi/6,  "BO"),
            (2, 3*math.pi/6,  "CO"),
            (2, 7*math.pi/6,  "CO"),
            (2, 11*math.pi/6, "CO"),
        ]:
        angle = R + offset
        pt = (center[0] + mult * x * math.cos(angle),
              center[1] + mult * x * math.sin(angle))
        pts.add((pt, label))
    return pts

def gen_points(k):
    """
    Generates points (each as ((x, y), label)) arranged in a spiral of triangles.
    Duplicate coordinates (after rounding) are filtered out.
    The inner loop has been rewritten to mimic the original offset logic.
    """
    points = set()
    seen = set()  # To avoid duplicate (rounded) coordinates
    i, increment, start = 0, 1, 0
    base = 0  # Horizontal base offset (s0 in the original)
    
    while True:
        if i == k - 1:
            increment = -1
        # Loop over positions in this row: j runs from 1 to i+1.
        for j in range(1, i + 2):
            # Original: for q in range(start+1, start+i+2) with an offset_counter that increments every even j.
            q = start + j
            offset = (j - 1) // 2
            # Compute radius following the original: 2 * x * (2*start + q + offset)
            radius = 2 * x * (2 * start + q + offset)
            center = (base + radius * math.cos(math.pi/6),
                      radius * math.sin(math.pi/6))
            # Alternate triangle orientation based on parity of j.
            tri = gen_triangle(center, reversed=(j % 2 == 0))
            for pt, label in tri:
                rounded = (round(pt[0], 2), round(pt[1], 2))
                if rounded not in seen:
                    seen.add(rounded)
                    points.add(((pt[0], pt[1]), label))
        base -= 2 * x * math.sqrt(3)
        i += increment
        if increment == -1:
            start += 1
        if i < 0:
            break
    return points

# Generate our points.
P = list(gen_points(36))
n = len(P)
total_triangles = math.comb(n, 3)

# Cache repeated axis comparisons.
axis_cache = {}
def axis_pair(i, j):
    key = (i, j)
    if key not in axis_cache:
        axis_cache[key] = are_same_axis(P[i], P[j])
    return axis_cache[key]

# Count valid triangles.
count = 0
for i, j, k in tqdm(combinations(range(n), 3), total=total_triangles):
    # Check non-degeneracy (via area) and that all three pairs share an allowed axis.
    if (not is_degenerate_triangle(P[i][0], P[j][0], P[k][0]) and
        axis_pair(i, j) and axis_pair(i, k) and axis_pair(j, k)):
        count += 1

print(count)

File: scripts/163.cc
#include <iostream>
#include <cmath>
#include <vector>
#include <set>
#include <string>
#include <tuple>
#include <unordered_map>
#include <sstream>
#include <iomanip>

// Use our own definition of PI.
const double PI = 3.14159265358979323846;
// Global constant.
const double x_const = 1.0;

// Round a double to 'decimals' places.
double roundTo(double value, int decimals) {
    double factor = std::pow(10.0, decimals);
    return std::round(value * factor) / factor;
}

// Rounded tangent: tan(angle) rounded to 3 decimals.
double tanRounded(double angle) {
    return roundTo(std::tan(angle), 3);
}

// A point is defined by its coordinates and a label.
struct Point {
    double x, y;
    std::string label;
};

// For use in sets to filter duplicate (rounded) coordinates.
struct Coord {
    double x, y;
    bool operator<(const Coord &other) const {
        if (std::fabs(x - other.x) > 1e-9)
            return x < other.x;
        return y < other.y;
    }
};

// Allowed slopes per label.
// For vertical lines we “store” a None in Python but here we simply omit it,
// since verticality is handled separately.
std::unordered_map<std::string, std::vector<double>> AXIS_SLOPES = {
    {"CO", {0.0, tanRounded(PI/6), tanRounded(PI/3), tanRounded(2*PI/3), tanRounded(5*PI/6)}},
    {"C",  {tanRounded(PI/6), tanRounded(5*PI/6)}},
    {"TR", {tanRounded(PI/6), tanRounded(2*PI/3)}},
    {"BO", {0.0}},
    {"TL", {tanRounded(5*PI/6), tanRounded(PI/3)}}
};

// are_same_axis: returns true if two points share an allowed axis.
// If the x–difference is nearly zero (vertical), we return true if b’s label is one
// of "CO", "C", or "BO". Otherwise we compute the slope and check against a's allowed slopes.
bool are_same_axis(const Point &a, const Point &b) {
    if (std::fabs(a.x - b.x) < 1e-3) {
        return (b.label == "CO" || b.label == "C" || b.label == "BO");
    }
    double slope = (a.y - b.y) / (a.x - b.x);
    double angle = std::atan(slope);
    double tan_val = roundTo(std::tan(angle), 3);
    auto it = AXIS_SLOPES.find(a.label);
    if (it != AXIS_SLOPES.end()) {
        for (double allowed : it->second) {
            if (std::fabs(allowed - tan_val) < 1e-3)
                return true;
        }
    }
    return false;
}

// is_degenerate_triangle: returns true if the triangle defined by points a, b, c is degenerate
// (i.e. two points coincide or the area is nearly zero).
bool is_degenerate_triangle(const Point &a, const Point &b, const Point &c, double tol = 1e-6) {
    if ((std::fabs(a.x - b.x) < 1e-9 && std::fabs(a.y - b.y) < 1e-9) ||
        (std::fabs(b.x - c.x) < 1e-9 && std::fabs(b.y - c.y) < 1e-9) ||
        (std::fabs(a.x - c.x) < 1e-9 && std::fabs(a.y - c.y) < 1e-9))
        return true;
    double area = std::fabs(a.x*(b.y - c.y) + b.x*(c.y - a.y) + c.x*(a.y - b.y)) / 2.0;
    return area < tol;
}

// gen_triangle: given a center (as a pair) and a flag 'reversed', returns a vector of Points
// corresponding to the triangle (including the center with label "C" and six surrounding points).
std::vector<Point> gen_triangle(const std::pair<double, double>& center, bool reversed) {
    std::vector<Point> pts;
    pts.push_back({center.first, center.second, "C"});
    double R = reversed ? PI : 0.0;
    // Each tuple: (multiplier, angle offset, label)
    std::vector<std::tuple<double, double, std::string>> defs = {
        {1, PI/6,    "TR"},
        {1, 5*PI/6,  "TL"},
        {1, 9*PI/6,  "BO"},
        {2, 3*PI/6,  "CO"},
        {2, 7*PI/6,  "CO"},
        {2, 11*PI/6, "CO"}
    };
    for (auto &def : defs) {
        double mult, offset;
        std::string label;
        std::tie(mult, offset, label) = def;
        double angle = R + offset;
        double pt_x = center.first + mult * x_const * std::cos(angle);
        double pt_y = center.second + mult * x_const * std::sin(angle);
        pts.push_back({pt_x, pt_y, label});
    }
    return pts;
}

// gen_points: generates points arranged in a spiral of triangles.
// Duplicate points (after rounding to 2 decimals) are filtered out.
std::vector<Point> gen_points(int k) {
    std::set<Coord> seen;
    std::vector<Point> points;
    int i = 0, increment = 1, start = 0;
    double base = 0.0;
    while (true) {
        if (i == k - 1)
            increment = -1;
        // Loop over positions in this row: j runs from 1 to i+1.
        for (int j = 1; j <= i + 1; j++) {
            int q = start + j;
            int offset = (j - 1) / 2;
            double radius = 2 * x_const * (2 * start + q + offset);
            double center_x = base + radius * std::cos(PI/6);
            double center_y = radius * std::sin(PI/6);
            std::pair<double, double> center = {center_x, center_y};
            bool reversed = (j % 2 == 0);
            std::vector<Point> tri = gen_triangle(center, reversed);
            for (auto &p : tri) {
                double rx = roundTo(p.x, 2);
                double ry = roundTo(p.y, 2);
                Coord rounded = {rx, ry};
                if (seen.find(rounded) == seen.end()) {
                    seen.insert(rounded);
                    points.push_back(p);
                }
            }
        }
        base -= 2 * x_const * std::sqrt(3);
        i += increment;
        if (increment == -1)
            start++;
        if (i < 0)
            break;
    }
    return points;
}

int main() {
    // Generate our points.
    std::vector<Point> P = gen_points(3);
    int n = P.size();

    // Precompute axis comparisons cache:
    // Create an n x n matrix (only i < j used) storing:
    //   -1: not computed; 0: false; 1: true.
    std::vector<std::vector<int>> axis_cache(n, std::vector<int>(n, -1));
    auto axis_pair = [&](int i, int j) -> bool {
        if (axis_cache[i][j] == -1) {
            bool res = are_same_axis(P[i], P[j]);
            axis_cache[i][j] = res ? 1 : 0;
        }
        return axis_cache[i][j] == 1;
    };

    // Total number of triangle combinations.
    unsigned long long total_combinations = (unsigned long long)n * (n - 1) * (n - 2) / 6;
    unsigned long long progressInterval = total_combinations / 100;
    if (progressInterval == 0) progressInterval = 1;

    // Count valid triangles.
    // A valid triangle is nondegenerate and all three pairs share an allowed axis.
    unsigned long long count = 0;
    unsigned long long iter = 0;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            for (int k = j + 1; k < n; k++) {
                iter++;
                if (iter % progressInterval == 0) {
                    double progress = (double)iter / total_combinations * 100.0;
                    std::cerr << "\rProgress: " << std::fixed << std::setprecision(6)
                              << progress << "% " << std::flush;
                }
                if (!is_degenerate_triangle(P[i], P[j], P[k]) &&
                    axis_pair(i, j) && axis_pair(i, k) && axis_pair(j, k))
                {
                    count++;
                }
            }
        }
    }
    std::cerr << "\rProgress: 100.00% " << std::endl;
    std::cout << count << std::endl;
    return 0;
}
